src/engine/enginebufferscalelinear.cpp

```diff
@@ -184,14 +183,12 @@ CSAMPLE* EngineBufferScaleLinear::do_scale(CSAMPLE* buf,
         CSAMPLE* write_buf = buf;
 
         // Use up what's left of the internal buffer.
-        int iNextSample = static_cast<int>(ceil(m_dNextSampleIndex)) * 2;
-        if (iNextSample + 1 < buffer_int_size) {
-            for (int i = iNextSample;
-                 samples_needed > 2 && i < buffer_int_size; i += 2) {
-                *write_buf = buffer_int[i]; write_buf++;
-                *write_buf = buffer_int[i+1]; write_buf++;
-                samples_needed -= 2;
-            }
+        int iNextSample = static_cast<int>(ceil(m_dNextFrame)) * 2;
+        int readSize = math_min<int>(m_bufferIntSize - iNextSample, samples_needed);
+        if (readSize > 0) {
+            SampleUtil::copy(write_buf, &m_bufferInt[iNextSample], readSize);
+            samples_needed -= readSize;
+            write_buf += readSize;
```

Comment:
nice refactoring!  I spent a long time staring at this and I think it's right :)